Performance monitoring

pwinston opened the pull request:
## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Description
This PR adds performance monitoring. We created performance monitoring to help in the investigation of #845, #1300 and #1320. However we want performance monitoring to be a permanent part of napari. There will always be performance issues cropping up that need to be investigated.

Care was taken so the performance monitoring code has zero impact on anything unless it is enabled with this `NAPARI_PERFMON` environment variable. We don't want regular users to take any sort of performance hit from the monitoring code. Although note that even when enabled performance monitoring is *much* lighter weight than comprehensive profiling.

The performance monitoring is based on timers than have a name and a duration. Timing data comes from 3 sources:

1. Our special `QApplication` that times all **Qt Events**.
1. Our`perf_timer` context object to time specific blocks of code.
1. Our `perf_func` decorator to time a specific functions.

All 3 types of timers call into `perf_counter_ns()` which itself takes about 100ns to run. However there are other costs to timers including writing the trace file.

The timing data is displayed in two ways:

1. Our QtPerformance widget:

<img width="629" alt="perf-window" src="https://user-images.githubusercontent.com/4163446/82953236-65bd0600-9f78-11ea-9d7e-feb33df81ac5.png">

Right now this only prints "slow events". It does not give a graphical picture of anything or show the hierarchy of events.

2. From the debug menu you can record "trace files" which can be viewed in Chrome using the special `chrome://tracing` URL. This shows **every** timing event that was recorded in a nice graphical view that you can pan/zoom.

Create a trace file from the Debug Menu:

<img width="520" alt="trace-menu" src="https://user-images.githubusercontent.com/4163446/82952125-66ed3380-9f76-11ea-9f74-5d9a6670f4ef.png">

Drag-n-drop the trace file into `chrome://tracing` to view:

<img width="645" alt="napari-long-paint" src="https://user-images.githubusercontent.com/4163446/82232537-c4410f00-98fc-11ea-9f4c-7052f3c8f0f6.png">

## How To Use

Typical investigation of a performance issue might go like this:

1. Run napari with `NAPARI_PERFMON=1`. This works from command line or from IPython/Jupyter.
2. View slow events in the QtPerformance widget.
3. Use the Debug menu to create a trace file, view it with `chrome://tracing`.
4. To dig in deeper add more timers by hand using our `perf_timer` and `perf_func` utilities. 

In theory `perf_timer` and `perf_func` calls can be kept in permanently. They should be 100% disabled and take no time if `NAPARI_PERFMON` is off. However I have not started doing that with this PR. I've been adding timers when exploring but then reverting them.

Note you *can* add perf timers to `vispy`. Just import  `napari.utils.perf as perf` it works even though it seems circular. It can be extremely helpful to see `napari` and `vispy` perf times combined together in one display.

# Trace Files

Chrome's `chrome://tracing` is a very nice way to view trace files. I used the `WASD` keys along with the mouse. The trace file format is JSON. Their file format has a lot features we aren't using yet.

You can also view the same trace files in:
* https://www.speedscope.app/
* https://doc.qt.io/qtcreator/creator-qml-performance-monitor.html

Speedscope can show a [flamegraph](http://www.brendangregg.com/flamegraphs.html) view that Chrome cannot. Speedscope calls the view "Left Heavy" instead of "flamegraph", but it's the same thing.

With flamegraphs the X axis is not time, it's more like "blame". If while timer A is running we make 50 separate calls into timer B, the flamegraph will show that A called B once, but the duration of the call will be the sum of those 50 times we called B. The benefit is it shows how expensive B was in total, which may not be clear in the view showing calls over time.

## Terminology
**Performance monitoring** is meant to be much lighter weight than profiling. Ideally performance timers can be left on at all time, or at least can be toggled on easily like we are doing. Profiling is great because it can time every single function at once, the downside is profiling can sometimes make interactive programs unusuably slow.

Both performance monitoring and profiling have their place.

## Overhead

Performance monitoring can slow things down. This is a form of the "observer effect". The root `perf_counter_ns()` call only takes around 100ns but other parts are probably a lot slower:

1. Updating the `Stat` objects in the `PerfTimers` dict.
1. Writing the JSON file when tracing is active.
1. Updating the QtPerformance widget.

We have not explored the overhead of timing yet in detail. That's one reason we are leaving it opt-in for now. Anecdotally the overhead seems low.

## How has this been tested?
- [x] Used a lot by hand to investigate performance issues.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas

pwinston committed:
Initial experimental performance monitoring

If NAPARI_PERFMON is defined we use our own QApplication
that times events and sends them to our new PerfTimers
class.

If NAPARI_PERFMON_TRACE_PATH is defined them PerfTimers
will write the timers to a chrome://tracing support log file.

NAPARI_PERFMON also creates a new dock widget QtPerformance
which is very much a WIP. It shows the average UpdateRequest time
which the main paint/draw event that hangs if we are doing IO
during the draw. It also has a text window where it "lots" any
"slow" events (UpdateRequest or otherwise).

pwinston committed:
Cleanup, remove unused code, remove print

Create LONG_EVENT_MS class var
Remove print and replace with comment.

pwinston committed:
Remove stale comment

pwinston committed:
Better comments

pwinston committed:
Improve comments

pwinston committed:
Set max/min not needed

pwinston committed:
perf_timers and tracing comments

Big comment on the lack of full nesting
support for PerfTimers

pwinston committed:
Define a perf_counter_ns() to support Python 3.6

pwinston committed:
Cleanup QtPerformance, add "uptime"

Add "draw time" and "slow events' labels.
Add lots of comments.
Extract method: _get_timer_info

pwinston committed:
Better comment describing the layout

pwinston committed:
Add perf_timer context object

pwinston committed:
Big iteration of perfmon stuff, put in own utils/perf folder (#1)

* A lot of cleanup/improvement to perf stuff
* Move USE_PERFMON check to event_loop.py
* Split things into mutiple files
* New Debug menu and trace file recording menu options.

pwinston committed:
Merge branch 'master' of github.com:napari/napari into feature/perf-mon

pwinston committed:
Remove stray method

pwinston committed:
use perf.USE_PERFOMON instead of env var

pwinston committed:
Add a combo box to set threashold (#2)

pwinston committed:
Fix decorator

pwinston committed:
Need category, placeholder for now

pwinston committed:
Merge branch 'master' into feature/perf-mon

pwinston committed:
New perfmon that works with Jupyter (#3)

New convert_app_for_timing works with Jupyter

The old approach was restricted to gui_qt which is only
used when running napari from the command line.

Now we have 2 approaches. We still do gui_qt for the
console app. But in Viewer we call convert_app_for_timing
and this works for Jupyter.

pwinston committed:
Fix comment

pwinston committed:
Better classname

pwinston committed:
Better comment about QApplication

pwinston committed:
Add event to string name mapping for PyQt5

In PySide2 the str() method gives you a nice string name like
'PySide2.QtCore.QEvent.Type.KeyPress' But in PyQt5 it return
a string/integer like '6' which is annoying.

This creates a lookup table on startup. This should be 100% find and
perform fine. So we use it for both PyQt5 and PySide2 to be consisting
so we work the same in both.

pwinston committed:
Better comment on convert_app_for_timing call

pwinston committed:
Edits for readability. Plus add sys.argv to the QApplication()

pwinston committed:
Merge in perf_func/perf_func_name from other branch.

Most of the time we can just use perf_func that uses the function name
by default. Most of the time that's what we want.

sofroniewn reviewed (approved):
This is beautiful @pwinston - lots of fun to play with both with your widget and the tracing view in chrome!

![perf_mon](https://user-images.githubusercontent.com/6531703/83978053-ab06ff00-a8b9-11ea-9fe2-9691333b6579.gif)

I think I have a good understanding of how the code uses to QApplication notify method to do the timing, and everything is clearly documented. For me this is good to merge.

@tlambert03 do you want to take one final look before merge. Juan said this was good to go in while he was out, and it's something that as it is off by default and mainly for developers right now we can continue to iterate on and improve as we use it

tlambert03 commented on napari/_qt/event_loop.py:
```diff
@@ -6,6 +6,25 @@
+    call convert_app_for_timing() in Viewer but that one is for Jupyter/IPython
+    while this one is for running from the command line.
+    """
+    if perf.USE_PERFMON:
```
might we want to just directly use `os.getenv` here and put the `from .qt_event_timing` import below the conditional, to avoid importing everything in the perf module when we aren't using it?
(I looked through, and it seems to be small/fast imports... but even still?)

pwinston replied:
Yes fixed this so perf is imported lazy. I suspect the import time is negligible but it seems reasonable to keep perf totally lazy.

tlambert03 commented on napari/_qt/qt_debug_menu.py:
```diff
@@ -0,0 +1,107 @@
+
+            # Schedule this to avoid bogus "MetaCall" event for the entire
+            # time the file dialog was up.
+            QTimer.singleShot(0, start_trace)
```
this feels like a hack, how were you previously connecting the `start_trace` function and showing the dialog?   Could you use the `fileSelected` signal?

pwinston replied:
Good idea but I just tried it, same problem you get a giant `MetaCall` event in the trace, covering the many seconds you had the dialog open. It must call that signal immediately in a nested way. This kind of ruins the trace file so we don't want it there.

This is a very specialized situation related to timing. When `_start_trace` is called `QApplicationWithTiming` has a timer running. If we turn on tracing immediately that super long timer goes into the trace file.

The very first thing I tried was build logic into PerfTimers to say "when you start a trace, don't really start tracing for a little bit" but that seemed messy. I think `singleShot` does exactly what we want, it's one line, it's a legal public method. Plus this is in the debug menu.

tlambert03 commented on napari/viewer.py:
```diff
@@ -5,11 +5,13 @@
 from ._qt.threading import wait_for_workers_to_quit, create_worker
 from .components import ViewerModel
 from . import __version__
+from .utils import perf
```
same questions about lazy imports in `event_loop.py` can we avoid importing any of that machinery if we don't need it?

pwinston replied:
I made used env var and made the import lazy here too.

tlambert03 commented on napari/viewer.py:
```diff
@@ -62,6 +64,13 @@ def __init__(
+        # the special one, and this will be a noop. When running inside IPython
+        # or Jupyter however this is where we switch to our QApplication.
+        if perf.USE_PERFMON:
+            app = convert_app_for_timing(app)
```
if we can do this here, do we need it in `event_loop.py` at all?  Or do you think it's still better to do it before instantiation if at all possible?

pwinston replied:
The one in Viewer does not work for `gui_qt` because we've already created a QWidget (the splash screen) so we can't delete QApplication, it will crash. But with IPython/Jupyter we have not created any Qt objects at that point.

We could call `convert_app_for_timing` in `gui_qt`. That is call `convert_app_for_timing` in both places. But that means we'd create the `QApplication` then **immediately** delete it which seems kind of silly, since it's just as easy to create the right one.

So right now:
Command Line: create the right QApplication from the start.
IPython/Jupyter: delete the existing QApplication and replace it.

Where `convert_app_for_timing` does **not** delete the QApplication if it's already the perfmon one.

tlambert03 replied:
oh fascinating!  So is that the distinction here?  For instance, if someone is using Jupyter and they create some other widget (outside of napari) before running napari with perfmon, will it crash there to?  (recognize that's basically never going to happen... just want to make sure I understand)

I'm definitely fine with the current approach, just wanted to understand

pwinston replied:
I think that's right. I think napari would delete the "shared" QApplication and it would crash. If it was not the first. Running multiple napari windows should be okay. Because the 2nd one will see it's already using QApplicationWithTiming.

Anyway this worked for timing #1320 Juan's notebook so that was the immediate goal. But this could bite someone if they just kept perfmon enabled, it would be a very mysterious crash for them, if they just opened some random notebook and it crash. Might take a long while to suspect perfmon.

Between this and the crash on exit it'd be **really** nice to find a completely new way of timing events. It'd be nice if perfmon were bulletproof. But I'm probably going to focus on #1320 itself.

Maybe there can be a V2 of perfmon at some point.

tlambert03 replied:
yeah that works for me.  as an internal tool, it's plenty useful as is

tlambert03 commented on napari/utils/perf/_trace_file.py:
```diff
@@ -0,0 +1,92 @@
+        ----------
+        path : str
+            Write the trace file to this path.
+        """
```
side note here:  I can't find the conversations where this has been discussed at the moment, but for better or worse, we follow the [numpy convention](https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring) in the napari codebase, which is to put the full docstring for the init method in `Parameters` section in the class docstring.

pwinston replied:
Didn't realize that was the convention. Fixed.

tlambert03 commented on napari/_qt/qt_viewer.py:
```diff
@@ -189,6 +193,21 @@ def __init__(self, viewer):
+            QtPerformance(),
+            name='performance',
+            area='bottom',
+            allowed_areas=['bottom'],
```
while playing with this a bit ... I kinda _wanted_ to put it on the right?  Any reason to restrict this here?

pwinston replied:
I just picked bottom very early on. Seems to work fine on right so I changed it to allow either. We can change default to right if people want to?

tlambert03 replied:
I think bottom is a good default.  I'd just remove the `allowed_areas` restriction

pwinston replied:
Makes sense. I removed it.

tlambert03 commented on napari/utils/perf/_stat.py:
```diff
@@ -0,0 +1,56 @@
+    """
+
+    def __init__(self, value: int):
+        """Create Stat with an initial value.
+
+        Parameters
+        ----------
+        value : int
+            Initial value.
+        """
```
see other comment about class docstrings

pwinston replied:
Fixed.

tlambert03 reviewed (commented):
The functionality is fantastic... love it @pwinston.

I do have one major concern at the moment, which is that I get this error everytime I quit the program.  Do you ever see this?  (If not, and you're on PySide2 ... that might be worth looking into?)

```pytb
WARNING: StopIteration

The above exception was the direct cause of the following exception:

SystemError: <class 'PyQt5.QtWidgets.QWidget'> returned a result with an error set
16:20:20 WARNING StopIteration

The above exception was the direct cause of the following exception:

SystemError: <class 'PyQt5.QtWidgets.QWidget'> returned a result with an error set
Abort trap: 6
/Users/talley/miniconda3/envs/napdev/lib/python3.8/multiprocessing/resource_tracker.py:216: 
UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at 
shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

I left some other comments throughout

tlambert03 commented on napari/_qt/qt_performance.py:
```diff
@@ -0,0 +1,174 @@
+        # We let the user set the "slow event" threshold.
+        self.thresh_ms = self.THRESH_DEFAULT
+        self.thresh_combo = QComboBox()
+        self.thresh_combo.addItems(self.THRESH_OPTIONS)
```
another trick worth knowing about... the [`QComboBox.addItem` method](https://doc.qt.io/qt-5/qcombobox.html#addItem) (but not `addItems`) accepts an optional _UserData_ value that lets you store arbitrary data with an item ... which you can then retrieve using [`currentData`](https://doc.qt.io/qt-5/qcombobox.html#currentData-prop).  This let's you avoid all the boilerplate of needing to convert back and forth between strings and floats

pwinston replied:
That's good to know about. This doesn't seem bad with string/float but I could see `UserData` being cleaner in some cases.

pwinston committed:
Talley review: moves/fixed Parameters for __init__

pwinston committed:
Talley review: allow docking to the right

I just picked "bottom" early on before I knew what was going to be in the
widget. Works fine on the right so yes allowing that here.

pwinston committed:
Talley review: use env var directly and late import

Avoid importing perf at all unless we are really using it.

pwinston committed:
Better comment

pwinston committed:
Talley review: remove allowed_areas entirely
